examples/migrations_example/test/drift/default/validation/v10_to_v11.dart

```diff
@@ -0,0 +1,17 @@
+import '../schemas/schema_v10.dart' as v10;
+import '../schemas/schema_v11.dart' as v11;
+
+/// Run `dart run drift_dev make-migrations --help` for more information
+
+final usersV10 = <v10.UsersData>[];
```

Comment:
I'm not too sure about generating this into separate files. It adds a lot of new files, and users are expected to modify them either way right?

I don't see a strong advantage of these files over just inlining them in the generated example tests? E.g. what if we created

```dart

test("default - migrate from v10 to v11", () async {
  final oldUsers = [...]
  final migratedUsers = [...]

  // Now invoke test utility here
});
```

Is the problem that we'd then have to patch test files whereas we can emit these files once and ignore them if they already exists?
Since we can't generate meaningful examples to test consistency automatically, I don't think this is really worth it. I think I prefer the approach previously used in the migrations example which:

1. Tests all combinations of schema versions without data integrity checks.
2. For the migrations that are critical because they alter tables, write custom tests verifying data integrity.

I agree with the utility in `drift_dev` to make step 2 easier, but it can't be fully automated and we shouldn't try to. Maybe we can generate a single test file as an example if it doesn't already exist and then encourage users to write their migration tests in that file? It could contain these tests (with inlined versioned data lists for data integrity checks) as an example so that users know how to write them.

Reply:
> I'm not too sure about generating this into separate files. It adds a lot of new files, and users are expected to modify them either way right?

They don't HAVE to do anything. It's completely optional.

> Is the problem that we'd then have to patch test files whereas we can emit these files once and ignore them if they already exists?

Exactly, I don't want to overwrite existing tests.

We could do this instead
This is generated once and never overwritten.

validate.dart
```dart
import "../schema_v1.dart" as v1;

final validation = Validate(
  v1toV2: (
    v1: V1Models(
      users: [v1.User(...)]
    ),
    v2: V2Models(
      users: [v2.User(...)]
    ),
  ),
  v2toV3: (
      v2: V2Models(
        users: [v2.User(...)]
      ),
      v3: V3Models(
        users: [v3.User(...)]
      ),
    )
);
```


> I agree with the utility in drift_dev to make step 2 easier, but it can't be fully automated and we shouldn't try to.

I heat your point, a random file with some lists feels weird. The above setup not as much, but still funny.

The Validate class would be implemented in another file and would be overwritten on each run:
```dart
class Validate{
  final ({V1Models v1,V2Models v2}) v1toV2; 
  final ({V2Models v2,V3Models v3}) v2toV3; 
  //...
}
```


The downside is that users need to write the import themselves, which is ugly.
```dart
import "../schema_v1.dart" as v1;
```

Reply:
My point is that most migrations likely don't need a test verifying data integrity. Most migrations add new tables and columns, users know that and they know that inserting rows at the start and checking them in the end is a lot of effort that is just not needed for those tests.
In my experience, "dangerous" migrations that change column constraints or types in a way that could affect existing data are not so common. So my suggestion is that we don't generate these validation utilities for every migration. Still, it's important that we inform users that verification tests are possible and sometimes really helpful.

So, I think we should generate a single test file (if it doesn't exist already) that:

1. Uses the for-loop pattern previously shown in the migration example to test every possible migration sequence (this likely doesn't need customization for users, these tests are super simple but can still catch issues reliably. Basically the first 20% of effort that does 80% of the work).
2. Then, for any migration we know of (can just be the first one), sets up another test that looks like

```dart
test('migration from v1 to v2 does not corrupt data', () async {
  // TODO: Consider writing these kinds of tests when altering tables in a way that might affect existing rows.
  // The default migration tests run with an empty schema, so it's a recommended practice to also test with
  // data for relevant migrations.
  final oldData = [];  // TODO: Add expected data at version 1 using v1.User()
  final expectedNewData = []; // TODO: Add expected data at version 2 using v2.User()

  await verifier.testWithDataIntegrity(
    // parameters filled out
  )
});
```

(where `User` is replaced with a shared data class name for the schema versions)

We don't have to re-generate that test or the expected data ever. We've taught users about the importance of testing with data, now we have to hope that they will actually follow that advice when needed. That's the best we can do either way since users need to fill in the rows. I don't see how the `validate.dart` or the other additional files add enough value to be worth the clutter. Even with the manual import management I think it's still the better solution.

Reply:
I completely agree!
This is a great idea.


> Basically the first 20% of effort that does 80% of the work

Solid Point

> Even with the manual import management I think it's still the better solution.

We could open the file and inject the imports to the top of the file.

Reply:
This is what I have now.
Modify this to your liking and that's what I'll do:



```dart
// ignore_for_file: unused_local_variable, unused_import
// GENERATED CODE, DO NOT EDIT BY HAND.
import 'package:drift/drift.dart';
import 'package:drift_dev/api/migrations.dart';
import 'package:migrations_example/database.dart';
import 'package:test/test.dart';
import 'generated/schema.dart';

import 'generated/schema_v1.dart' as v1;
import 'generated/schema_v2.dart' as v2;

void main() {
  driftRuntimeOptions.dontWarnAboutMultipleDatabases = true;
  late SchemaVerifier verifier;

  setUpAll(() {
    verifier = SchemaVerifier(GeneratedHelper());
  });

  group('default database', () {
    //////////////////////////////////////////////////////////////////////////////
    ////////////////////// GENERATED TESTS - DO NOT MODIFY ///////////////////////
    //////////////////////////////////////////////////////////////////////////////
    if (GeneratedHelper.versions.length < 2) return;
    for (var i
        in List.generate(GeneratedHelper.versions.length - 1, (i) => i)) {
      final oldVersion = GeneratedHelper.versions.elementAt(i);
      final newVersion = GeneratedHelper.versions.elementAt(i + 1);
      test("migrate from v$oldVersion to v$newVersion", () async {
        final schema = await verifier.schemaAt(oldVersion);
        final db = Database(schema.newConnection());
        await verifier.migrateAndValidate(db, newVersion);
        await db.close();
      });
    }
    //////////////////////////////////////////////////////////////////////////////
    /////////////////////// END OF GENERATED TESTS ///////////////////////////////
    //////////////////////////////////////////////////////////////////////////////

    //////////////////////////////////////////////////////////////////////////////
    ///////////////////// CUSTOM TESTS - MODIFY AS NEEDED ////////////////////////
    //////////////////////////////////////////////////////////////////////////////
    test("migration from v1 to v2 does not corrupt data", () async {
      // TODO: Consider writing these kinds of tests when altering tables in a way that might affect existing rows.
      // The automatically generated migration tests run with an empty schema, so it's a recommended practice to also test with
      // data for relevant migrations.
      final oldUsersData = <v1
          .UsersData>[]; // TODO: Add expected data at version 1 using v1.UsersData
      final expectedNewUsersData = <v2
          .UsersData>[]; // TODO: Add expected data at version 2 using v2.UsersData

      await verifier.testWithDataIntegrity(
        oldVersion: 1,
        newVersion: 2,
        verifier: verifier,
        createOld: (e) => v1.DatabaseAtV1(e),
        createNew: (e) => v2.DatabaseAtV2(e),
        openTestedDatabase: (e) => Database(e),
        createItems: (batch, oldDb) {
          batch.insertAll(oldDb.users, oldUsersData);
        },
        validateItems: (newDb) async {
          expect(expectedNewUsersData, await newDb.select(newDb.users).get());
        },
      );
    });
    ///////////////////////////////////////////////////////////////////////////////
    /////////////////////// END OF CUSTOM TESTS ///////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////
  });
}

```

Reply:
Looks good, I've reworded some of the comments to make them fit into 80 char lines, but that template is exactly what I had in mind :+1: